### PR TITLE
fix validation errors and enable additional checks in CI

### PR DESCRIPTION
**Pull request by ibrokethecloud**
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

**Problem:**
Include additional ci checks and fix reconcile issues with device plugins.

Related to: https://github.com/harvester/harvester/issues/3524

**Solution:**

**Related Issue:**

**Test plan:**

### CODE REVIEW

File: pkg/controller/pcideviceclaim/pcideviceclaim_controller.go [hunk 1]
```diff
@@ -107,6 +108,8 @@ func (h *Handler) OnRemove(name string, pdc *v1beta1.PCIDeviceClaim) (*v1beta1.P
 		return pdc, err
 	}
 
+	lock.Lock()
```
Comment (guangbochen) [inline pkg/controller/pcideviceclaim/pcideviceclaim_controller.go:111]: any reasons why the lock is requried on the remove method?

File: pkg/controller/pcideviceclaim/pcideviceclaim_controller.go [hunk 2]
```diff
@@ -290,15 +287,27 @@ func (h *Handler) reconcilePCIDeviceClaims(name string, pdc *v1beta1.PCIDeviceCl
...
```
Comment (guangbochen) [inline pkg/controller/pcideviceclaim/pcideviceclaim_controller.go:290]: ditto

File: pkg/deviceplugins/device_manager.go [hunk 3]
```diff
@@ -148,91 +148,95 @@ func constructDPIdevices(pciDevices []*PCIDevice, iommuToPCIMap map[string]strin
...
-	sock, err := net.Listen("unix", dpi.socketPath)
+	sock, err := net.Listen("unix", dp.socketPath)
 	if err != nil {
 		return fmt.Errorf("error creating GRPC server socket: %v", err)
 	}
 
-	dpi.server = grpc.NewServer([]grpc.ServerOption{}...)
-	defer dpi.stopDevicePlugin()
+	dp.server = grpc.NewServer([]grpc.ServerOption{}...)
+	//defer dpi.stopDevicePlugin()
```
Comment (guangbochen) [inline pkg/deviceplugins/device_manager.go:184]: remove this line if not required.

File: pkg/deviceplugins/common.go [hunk 4]
```diff
@@ -127,20 +129,17 @@ func waitForGRPCServer(socketPath string, timeout time.Duration) error {
 
 // dial establishes the gRPC communication with the registered device plugin.
 func gRPCConnect(socketPath string, timeout time.Duration) (*grpc.ClientConn, error) {
-	c, err := grpc.Dial(socketPath,
-		grpc.WithInsecure(),
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
```
Comment (guangbochen) [inline pkg/deviceplugins/common.go:132]: any preference to initialize a new context rather than pass through the main ctx via the handler?

**Review by guangbochen — COMMENTED** [review]
btw, prefer to bump to golang 1.19 if it is a manageable change, thanks.

File: pkg/controller/pcideviceclaim/pcideviceclaim_controller.go (hunk 1 above)
Comment (ibrokethecloud) [reply pkg/controller/pcideviceclaim/pcideviceclaim_controller.go:111]: The controller is running two threads by default.

The locking was introduced to ensure that device plugin updates are serialized.

The remove is also managed by a lock, as the plugin is deleted if there are no more pcideviceclaims for the device type and we need to ensure that plugin is working until the last deviceclaim is deleted.

File: pkg/controller/pcideviceclaim/pcideviceclaim_controller.go (hunk 2 above)
Comment (ibrokethecloud) [reply pkg/controller/pcideviceclaim/pcideviceclaim_controller.go:290]: same as above..

**Comment by ibrokethecloud** [comment]
To test device plugin fix can only be tested with multiple devices of the same type.

Easiest option would be to create VF's on a nic or use the SriovNetworkDevices crd.

* create 4 VF's
* enable pcidevice passthrough for all 4 VF's.
* check node status `kubectl get nodes -o yaml | yq .status.allocatable`, user should see 4 devices available for the VF's just created.
* delete pcideviceclaim pod or reboot node.
* check node status `kubectl get nodes -o yaml | yq .status.allocatable`, without this fix there is a high chance of only seeing 3 devices available for the VF's created from step above.
* build a new image with fixes, and update pcidevices-controller
* once controller initialises `kubectl get nodes -o yaml | yq .status.allocatable` should reflect the right number of VF's.

File: pkg/controller/pcideviceclaim/pcideviceclaim_controller.go (hunk 1 above)
Comment (guangbochen) [reply pkg/controller/pcideviceclaim/pcideviceclaim_controller.go:111]: Adding this as a code comment will be helpful, thanks.

File: pkg/deviceplugins/common.go (hunk 4 above)
Comment (ibrokethecloud) [reply pkg/deviceplugins/common.go]: logic has been updated to store the main context into pcideviceclaim handler, and this is in turn stored in pcideviceplugin, for use when the grpc connection is initialized.

File: pkg/controller/pcideviceclaim/pcideviceclaim_controller.go (hunk 1 above)
Comment (ibrokethecloud) [reply pkg/controller/pcideviceclaim/pcideviceclaim_controller.go:123]: done.

**Review by guangbochen — APPROVED** [review]
LGTM, thanks.

File: pkg/webhook/vm.go [hunk 5]
```diff
@@ -88,7 +88,7 @@ func (vm *vmPCIMutator) Update(request *types.Request, oldObj runtime.Object, ne
 
 // generatePatch is a common method used by create and update calls to generate a patch operation for VM
 func (vm *vmPCIMutator) generatePatch(vmObj *kubevirtv1.VirtualMachine) (types.PatchOps, error) {
-	var pciDevicesInVM []string
+	pciDevicesInVM := make([]string, len(vmObj.Spec.Template.Spec.Domain.Devices.HostDevices))
```
Comment (futuretea) [inline pkg/webhook/vm.go:91]: pciDevicesInVM := make([]string, 0, len(vmObj.Spec.Template.Spec.Domain.Devices.HostDevices))

Comment (ibrokethecloud) [reply pkg/webhook/vm.go]: fixed

**Review by futuretea — APPROVED** [review]
LGTM, thanks.

**Review by bk201 — APPROVED** [review]

**Comment by WebberHuang1118** [comment]
https://github.com/ibrokethecloud/pcidevices/blob/b6653138864acb5fe4842f2d0864eb71d6c0da2e/pkg/controller/pcidevice/pcidevice_controller.go#L69
maybe not directly related to this PR, but it logs with wrong information?

**Comment by ibrokethecloud** [comment]
good spot @WebberHuang1118. fixed it.

**Review by WebberHuang1118 — APPROVED** [review]
LGTM, thanks

**Pull request merged**